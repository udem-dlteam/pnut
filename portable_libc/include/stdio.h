#ifndef _STDIO_H
#define _STDIO_H

#include "sys/types.h"
#include "stdarg.h"

#ifdef USE_STRUCT

typedef struct {
  int fd;
  char buf[1];
  char *string_output_buf;
  size_t string_output_buf_size;
  size_t string_output_len;
} FILE;

#else

typedef int FILE;

#endif

#ifdef PNUT_CC

FILE *stdin;
FILE *stdout;
FILE *stderr;

#else

extern FILE *stdin;
extern FILE *stdout;
extern FILE *stderr;

#endif

FILE *fopen(const char *pathname, const char *mode);
FILE *fdopen(int fd, const char *mode);
int fclose(FILE *stream);

int fputc(int c, FILE *stream);
size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream);
int fputs(const char *s, FILE *stream);
int puts(const char *s);

size_t fread(void *buffer, size_t size, size_t count, FILE *stream);
int fseek( FILE* stream, long offset, int origin );
long ftell( FILE* stream );
int remove(const char *_Filename);

int vfprintf(FILE *stream, const char *format, va_list ap);
int fprintf(FILE *stream, const char *format VAR_ARGS);
int printf(const char *format VAR_ARGS);

int vsnprintf(char *str, size_t size, const char *format, va_list ap);
int snprintf(char *str, size_t size, const char *format VAR_ARGS);
int sprintf(char *str, const char *format VAR_ARGS);

#endif
