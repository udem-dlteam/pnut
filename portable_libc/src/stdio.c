#include "../include/stdio.h"

#include "../include/stdlib.h"
#include "../include/unistd.h"
#include "../include/fcntl.h"

FILE _standard_files[4];

FILE *stdin = _standard_files+0;
FILE *stdout = _standard_files+1;
FILE *stderr = _standard_files+2;

char *_string_output_buf;
size_t _string_output_buf_size;
size_t _string_output_len;

int _get_fd(FILE *f) {
  if (f == _standard_files+0) {
    return 0;
  } else if (f == _standard_files+1) {
    return 1;
  } else if (f == _standard_files+2) {
    return 2;
  } else {
    return *f;
  }
}

int fopen_flags(const char *mode) {
  int res = 0;
  while (*mode == 'b') ++mode; // Ignore binary mode

  if (*mode == 'r') {
    res = O_RDONLY;
  } else if (*mode == 'w') {
    res = O_WRONLY | O_CREAT | O_TRUNC;
  } else if (*mode == 'a') {
    res = O_WRONLY | O_CREAT | O_APPEND;
  } else {
    return 0;
  }

  mode += 1;
  while (*mode == 'b') ++mode; // Ignore binary mode (it can appear after r, w, a)

  if (*mode == '+') {
    res = res & ~(O_RDONLY | O_WRONLY) | O_RDWR;
  }

  return res;
}

FILE *fdopen(int fd, const char *mode) {
  FILE *result = malloc(sizeof(FILE));
  if (result) {
    *result = fd;
  }
  return result;
}

FILE *fopen(const char *pathname, const char *mode) {
  int fd = open(pathname, fopen_flags(mode), 0666); // 0666 is the default mode
  if (fd == -1) {
    return 0;
  } else {
    return fdopen(fd, mode);
  }
}

int fclose(FILE *stream) {
  int fd = _get_fd(stream);
  free(stream);
  if (close(fd) == 0) {
    return 0;
  } else {
    return EOF;
  }
}

int fseek(FILE* stream, long offset, int origin) {
  return lseek(_get_fd(stream), offset, origin);
}

off_t ftell(FILE * stream) {
  return lseek (_get_fd(stream), 0, SEEK_CUR);
}

size_t fread(void *data, size_t size, size_t count, FILE *stream) {
  return read(_get_fd(stream), data, size * count) / size;
}

int remove(const char *filename) {
  return unlink(filename);
}

char _output_buf[1];

int fputc(int c, FILE *stream) {
  int fd = _get_fd(stream);
  if (fd == -1) {
    if (_string_output_len+1 < _string_output_buf_size) {
      _string_output_buf[_string_output_len] = c;
      ++_string_output_len;
    }
  } else {
    _output_buf[0] = c;
    write(fd, _output_buf, 1);
  }
  return c;
}

size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream) {
  const char *p = ptr;
  size_t n = size * nmemb;
  while (n) {
    fputc(*p, stream);
    ++p;
    --n;
  }
  return nmemb;
}

int _fputstr(const char *str, FILE *stream) {
  int result = 0;
  while (*str) {
    fputc(*str, stream);
    ++str;
    ++result;
  }
  return result;
}

int fputs(const char *s, FILE *stream) {
  _fputstr(s, stream);
  // fputc('\n', stream);
  return 0;
}

int puts(const char *s) {
  return fputs(s, stdout);
}

int fflush(FILE *stream) {
  return 0; // no buffering so nothing to do
}

#define SIZEOF_NUM_BUF 100
char _num_buf[SIZEOF_NUM_BUF];

char *_int_to_str(int n, int base, int width, int force_0, int force_plus) {

  char *out = _num_buf + SIZEOF_NUM_BUF;
  int neg = n < 0;
  int has_sign = neg || force_plus;
  char *digits = "0123456789abcdef";

  if (!neg) n = -n;

  *--out = 0;

  while (n <= -base) {
    *--out = digits[-(n % base)];
    n /= base;
  }

  *--out = digits[-(n % base)];

  if (width >= SIZEOF_NUM_BUF) width = SIZEOF_NUM_BUF-1;

  if (force_0) {
    while (_num_buf + SIZEOF_NUM_BUF - out <= width - has_sign) {
      *--out = force_0 ? '0' : ' ';
    }
  }

  if (has_sign) *--out = neg ? '-' : '+';

  if (!force_0) {
    while (_num_buf + SIZEOF_NUM_BUF - out <= width) {
      *--out = ' ';
    }
  }

  return out;
}

int vfprintf(FILE *stream, const char *format, va_list ap) {

  char c;
  int result = 0;
  int base;
  int width;
  int force_0;

  while ((c = *format++)) {
    if (c == '%' && (c = *format++) != '%') {
      width = 0;
      force_0 = 0;
      if (c == '0') {
        force_0 = 1;
        c = *format++;
      }
      while (c >= '0' && c <= '9') {
        width = width*10 + (c - '0');
        c = *format++;
      }
      while (c == 'l' || c == 'L') {
        c = *format++;
      }
      if (c == 'c') {
        fputc(va_arg(ap, int), stream);
        result += 1;
      } else if (c == 'd' || c == 'u' || c == 'o' || c == 'x') {
        base = c == 'x' ? 16 : c == 'o' ? 8 : 10;
        result += _fputstr(_int_to_str(va_arg(ap, int), base, width, force_0, 0), stream);
      } else if (c == 's') {
        result += _fputstr(va_arg(ap, char*), stream);
      } else {
        fputc('?', stream);
        fputc('?', stream);
        fputc('?', stream);
        result += 3;
      }
    } else {
      fputc(c, stream);
      result += 1;
    }
  }

  return result;
}

int fprintf(FILE *stream, const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start(ap, format);
  result = vfprintf(stream, format, ap);
  va_end(ap);

  return result;
}

int printf(const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start(ap, format);
  result = vfprintf(stdout, format, ap);
  va_end(ap);

  return result;
}

int vsnprintf(char *str, size_t size, const char *format, va_list ap) {

  int result;

  _standard_files[3] = -1; /* mark as string FILE */
  _string_output_buf = str;
  _string_output_buf_size = size;
  _string_output_len = 0;
  result = vfprintf(_standard_files+3, format, ap);
  _string_output_buf[_string_output_len] = 0; /* null terminate string */

  return result;
}

int snprintf(char *str, size_t size, const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start(ap, format);
  result = vsnprintf(str, size, format, ap);
  va_end(ap);

  return result;
}

int sprintf(char *str, const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start(ap, format);
  result = vsnprintf(str, 999999999, format, ap);
  va_end(ap);

  return result;
}
