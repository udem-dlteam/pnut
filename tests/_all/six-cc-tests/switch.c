#include <stdio.h>

#ifdef PNUT_CC
typedef int FILE;
#endif

enum Category {
  LOWERCASE = 1,
  UPPERCASE,
  NEWLINE,
  OTHER
};

/* For all ascii characters, categorize them into 4 categories */
int categorize_char(char c) {
  int a;
  switch (c) {
    case 'a':
    case 'b':
    case 'c':
    case 'd':
    case 'e':
    case 'f':
    case 'g':
    case 'h':
    case 'i':
    case 'j':
    case 'k':
    case 'l':
    case 'm':
    case 'n':
    case 'o':
    case 'p':
    case 'q':
    case 'r':
    case 's':
    case 't':
    case 'u':
    case 'v':
    case 'w':
    case 'x':
    case 'y':
    case 'z':
      a = LOWERCASE;
      return LOWERCASE;
    case 'A':
    case 'B':
    case 'C':
    case 'D':
    case 'E':
    case 'F':
    case 'G':
    case 'H':
    case 'I':
    case 'J':
    case 'K':
    case 'L':
    case 'M':
    case 'N':
    case 'O':
    case 'P':
    case 'Q':
    case 'R':
    case 'S':
    case 'T':
    case 'U':
    case 'V':
    case 'W':
    case 'X':
    case 'Y':
    case 'Z':
      return UPPERCASE;
    case '\n':
      return NEWLINE;
    default:
      return OTHER;
  }
  /* Dead code, but puts the switch in non-tail call position */
  return a;
}

void putstring(char *s) {
  while (*s) {
    putchar(*s);
    s = s + 1;
  }
}

int main() {
  FILE *f;
  char c;
  f = fopen("tests/_all/six-cc-tests/fgetc.c", "r");
  while ((c = fgetc(f)) != -1) {
    putstring("'");
    putchar(c);
    putstring("' = ");
    putchar(c);
    putstring(": ");
    putchar(categorize_char(c) + 48);
    putchar('\n');
  }

  return 0;
}
