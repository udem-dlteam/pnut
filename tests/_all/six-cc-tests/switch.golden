'#' = #: 4
'i' = i: 1
'n' = n: 1
'c' = c: 1
'l' = l: 1
'u' = u: 1
'd' = d: 1
'e' = e: 1
' ' =  : 4
'<' = <: 4
's' = s: 1
't' = t: 1
'd' = d: 1
'i' = i: 1
'o' = o: 1
'.' = .: 4
'h' = h: 1
'>' = >: 4
'
' = 
: 3
'
' = 
: 3
'#' = #: 4
'i' = i: 1
'f' = f: 1
'd' = d: 1
'e' = e: 1
'f' = f: 1
' ' =  : 4
'P' = P: 2
'N' = N: 2
'U' = U: 2
'T' = T: 2
'_' = _: 4
'C' = C: 2
'C' = C: 2
'
' = 
: 3
't' = t: 1
'y' = y: 1
'p' = p: 1
'e' = e: 1
'd' = d: 1
'e' = e: 1
'f' = f: 1
' ' =  : 4
'i' = i: 1
'n' = n: 1
't' = t: 1
' ' =  : 4
'F' = F: 2
'I' = I: 2
'L' = L: 2
'E' = E: 2
';' = ;: 4
'
' = 
: 3
'#' = #: 4
'e' = e: 1
'n' = n: 1
'd' = d: 1
'i' = i: 1
'f' = f: 1
'
' = 
: 3
'
' = 
: 3
'i' = i: 1
'n' = n: 1
't' = t: 1
' ' =  : 4
'm' = m: 1
'a' = a: 1
'i' = i: 1
'n' = n: 1
'(' = (: 4
')' = ): 4
' ' =  : 4
'{' = {: 4
'
' = 
: 3
' ' =  : 4
' ' =  : 4
'F' = F: 2
'I' = I: 2
'L' = L: 2
'E' = E: 2
' ' =  : 4
'*' = *: 4
'f' = f: 1
';' = ;: 4
'
' = 
: 3
' ' =  : 4
' ' =  : 4
'c' = c: 1
'h' = h: 1
'a' = a: 1
'r' = r: 1
' ' =  : 4
'c' = c: 1
';' = ;: 4
'
' = 
: 3
' ' =  : 4
' ' =  : 4
'f' = f: 1
' ' =  : 4
'=' = =: 4
' ' =  : 4
'f' = f: 1
'o' = o: 1
'p' = p: 1
'e' = e: 1
'n' = n: 1
'(' = (: 4
'"' = ": 4
't' = t: 1
'e' = e: 1
's' = s: 1
't' = t: 1
's' = s: 1
'/' = /: 4
'_' = _: 4
'a' = a: 1
'l' = l: 1
'l' = l: 1
'/' = /: 4
's' = s: 1
'i' = i: 1
'x' = x: 1
'-' = -: 4
'c' = c: 1
'c' = c: 1
'-' = -: 4
't' = t: 1
'e' = e: 1
's' = s: 1
't' = t: 1
's' = s: 1
'/' = /: 4
'f' = f: 1
'g' = g: 1
'e' = e: 1
't' = t: 1
'c' = c: 1
'.' = .: 4
'c' = c: 1
'"' = ": 4
',' = ,: 4
' ' =  : 4
'"' = ": 4
'r' = r: 1
'"' = ": 4
')' = ): 4
';' = ;: 4
'
' = 
: 3
' ' =  : 4
' ' =  : 4
'w' = w: 1
'h' = h: 1
'i' = i: 1
'l' = l: 1
'e' = e: 1
' ' =  : 4
'(' = (: 4
'(' = (: 4
'c' = c: 1
' ' =  : 4
'=' = =: 4
' ' =  : 4
'f' = f: 1
'g' = g: 1
'e' = e: 1
't' = t: 1
'c' = c: 1
'(' = (: 4
'f' = f: 1
')' = ): 4
')' = ): 4
' ' =  : 4
'!' = !: 4
'=' = =: 4
' ' =  : 4
'-' = -: 4
'1' = 1: 4
')' = ): 4
' ' =  : 4
'{' = {: 4
'
' = 
: 3
' ' =  : 4
' ' =  : 4
' ' =  : 4
' ' =  : 4
'p' = p: 1
'u' = u: 1
't' = t: 1
'c' = c: 1
'h' = h: 1
'a' = a: 1
'r' = r: 1
'(' = (: 4
'c' = c: 1
')' = ): 4
';' = ;: 4
'
' = 
: 3
' ' =  : 4
' ' =  : 4
'}' = }: 4
'
' = 
: 3
' ' =  : 4
' ' =  : 4
'f' = f: 1
'c' = c: 1
'l' = l: 1
'o' = o: 1
's' = s: 1
'e' = e: 1
'(' = (: 4
'f' = f: 1
')' = ): 4
';' = ;: 4
'
' = 
: 3
'
' = 
: 3
' ' =  : 4
' ' =  : 4
'r' = r: 1
'e' = e: 1
't' = t: 1
'u' = u: 1
'r' = r: 1
'n' = n: 1
' ' =  : 4
'0' = 0: 4
';' = ;: 4
'
' = 
: 3
'}' = }: 4
'
' = 
: 3
